## Preliminary operations

In [1]:
# import main libraries
#import folium
import requests
import json
#import geopandas as geo
#fom io import StringIO
import io
import pandas as pd
from ortools.linear_solver import pywraplp
from numpy import average, std
from statistics import *
import folium
import numpy as np
from sklearn.preprocessing import MinMaxScaler

In [2]:
# degine vehicle types
vehicle_types = ["engine", "ladder", "rescue", "squad", "hazardous"]

In [3]:
# define serviceability distances (from neighborhoods_serviceability.ipynb)
serviceability_distances = {
    "engine": 800.24,
    "ladder": 800.24,
    "rescue": 2675.74,
    "squad": 2339.26,
    "hazardous": 4537.84,
}

In [4]:
# get distances between stations
response = requests.get("https://gitlab.com/drvicsana/cop-proyecto-2023/-/raw/main/project_data/distancias_estaciones.json")
distances_stations_db = json.loads(response.text)

# convert json to dataframe
df_distances_stations = pd.DataFrame(distances_stations_db)
# solve an error in the columns order
df_distances_stations =  pd.concat([df_distances_stations.iloc[-1:], df_distances_stations.iloc[:-1]], axis=0)
df_distances_stations.fillna(0, inplace=True)
df_distances_stations.head()

42 South Street  49 Beekman Street  100 Duane Street  \
42 South Street                0.00             228.56            282.07   
49 Beekman Street            154.94               0.00            175.70   
100 Duane Street             299.43             180.79              0.00   
14 N. Moore Street           408.12             289.48            174.89   
75 Canal Street              296.75             295.70            275.65   

                    14 N. Moore Street  75 Canal Street  25 Pitt Street  \
42 South Street                 367.00           299.82          397.01   
49 Beekman Street               260.63           261.52          391.57   
100 Duane Street                 96.76           272.78          413.52   
14 N. Moore Street                0.00           339.69          444.26   
75 Canal Street                 283.51             0.00          195.55   

                    222 East 2 Street  340 East 14 Street  \
42 South Street                523.64              609.91   
49 Beekman Street              494.09              549.25   
100 Duane Street               467.01              447.83   
14 N. Moore Street             450.79              431.61   
75 Canal Street                260.18              346.45   

                    253 Lafayette Street  42 Great Jones Street  ...  \
42 South Street                   444.76                 511.11  ...   
49 Beekman Street                 330.31                 403.98  ...   
100 Duane Street                  228.89                 302.56  ...   
14 N. Moore Street                212.68                 286.35  ...   
75 Canal Street                   260.93                 293.16  ...   

                    256 Hylan Blvd.  278 McClean Avenue  1850 Clove Road  \
42 South Street             1384.71             1348.97          1337.57   
49 Beekman Street           1401.76             1366.02          1354.62   
100 Duane Street            1411.62             1375.88          1364.48   
14 N. Moore Street          1520.31             1484.57          1473.17   
75 Canal Street             1563.49             1522.61          1516.35   

                    1592 Richmond Road  256 Nelson Ave.  345 Annadale Road  \
42 South Street                1553.31          2392.53            2406.78   
49 Beekman Street              1570.37          2409.58            2423.84   
100 Duane Street               1580.22          2419.44            2341.17   
14 N. Moore Street             1688.91          2405.59            2246.56   
75 Canal Street                1732.10          2571.31            2470.82   

                    1560 Drumgoole Road West  7219 Amboy Road  \
42 South Street                      2697.62          2833.72   
49 Beekman Street                    2609.88          2719.26   
100 Duane Street                     2469.12          2578.51   
14 N. Moore Street                   2374.51          2483.90   
75 Canal Street                      2598.77          2708.16   

                    1100 Rossville Ave  305 Front Street  
42 South Street                2500.30           1567.00  
49 Beekman Street              2396.89           1584.06  
100 Duane Street               2256.14           1593.91  
14 N. Moore Street             2161.52           1702.60  
75 Canal Street                2385.79           1745.79  

[5 rows x 219 columns]

In [5]:
# get ntas-station distances dataset
distance = pd.read_csv("dataset/distance_stations_neighborhoods.csv", index_col=0)
# get standardized neighborhoods info dataset
neighborhood = pd.read_csv("dataset/neighborhoods_info_standardized.csv")
# define a serviceability dictionary
serviceability = {}

# get the serviceability table for each vehicle type
for vehicle_type in vehicle_types:
    serviceability[vehicle_type] = pd.read_csv("dataset/serviceability/" + vehicle_type +"_ntas_serviceability.csv", index_col=0)

## Creating an instance of the Solver class

In [5]:
# define the solver type
solver = pywraplp.Solver.CreateSolver("SCIP")
# define a time limit for the solver (30 min)
solver.set_time_limit(60*1000*30)

## Setting parameters

- type of vehicles$: i=1\dots 5$

- stations: $j=1\dots 219$

- shifts: $k=1\dots 2$

- neighborhoods: $n=1 \dots 195$

In [7]:
response = requests.get("https://gitlab.com/drvicsana/cop-proyecto-2023/-/raw/main/project_data/incidentes2019.json")
incidents = json.loads(response.text)

I = ["engine", "ladder", "rescue", "squad", "hazardous"]
N = list(neighborhood.nta)
J = list(distance.columns)
K = [1,2]

## Creating decision variables

$X_{i, j, k}: $  number of vehicles of type ${i}$ deployed in station ${j}$ during shift ${k}$

$X_{i, j, k} \geq 0$, $X_{i, j, k} \in \mathbb{Z}$

$ i \in I, j \in J, k \in K$

In [9]:
X = {}
for i in I:
    for j in J:
        for k in K:
            v = solver.IntVar(0, solver.infinity(), "Number of vehicles of type " + str(i) + " deployed at station "+ str(j)+ " during shift " + str(k))
            X[(i,j,k)] = v

$Y_{i, a, b}:$  number of vehicles of type ${i}$ moved from station ${a}$ to station ${b}$ at shift change

$Y_{i, a, b} \geq 0$, $Y_{i, a, b} \in \mathbb{Z}$

$ i \in I, a \in J, b \in J$

In [10]:
Y = {}
for i in I:
    for a in J:
        for b in J:
            v = solver.IntVar(0, solver.infinity(), "Number of vehicles of type " + str(i) + " moved from station "+ str(a)+ " to station " + str(b))
            Y[(i,a,b)] = v

$T_{i, j, k}: 1$ if a vehicle of type ${i}$ is deployed in station ${j}$ during shift ${k}$, $0$ otherwise

$T_{i, j, k} \geq 0$, $T_{i, j, k} \in  \{0,1\}$

$ i \in I, j \in J, k \in K$

In [11]:
T = {}
for i in I:
    for j in J:
        for k in K:
            v = solver.BoolVar("if a vehicle of type" + str(i)  + " is deployed in station " + str(j) + " during shift " + str(k))
            T[(i,j,k)] = v

$Q_{j, k}: 1$ if there are more than two vehicles in station $j$ during shift $k$, $0$ otherwise

$Q_{j, k} \geq 0$, $Q_{j, k} \in \{0,1\}$

$ j \in J, k \in K$

In [12]:
Q = {}
for j in J:
    for k in K:
        v = solver.BoolVar("if there are two or more vehicles in station " + str(j) + " during shift " + str(k))
        Q[(j,k)] = v

## Creating constraints

### 1. Linking constraints

#### 1.1. Linking Variable T

$T_{i, j, k} \leq X_{i, j, k} \leq c_{j}T_{i, j, k}, \forall i \in I, j \in J, k \in K$

In [24]:
T_lower_linking_constraints = []
T_upper_linking_constraints = []

for i in I:
    for j in J:
        capacity = capacities[j]
        for k in K:
            v_T = T[(i, j, k)]
            v_X = X[(i, j, k)]

            c_lower = solver.Constraint(-solver.infinity(), 0)

            c_lower.SetCoefficient(v_T, 1)
            c_lower.SetCoefficient(v_X, -1)

            T_lower_linking_constraints.append(c_lower)

            c_upper = solver.Constraint(-solver.infinity(), 0)

            c_upper.SetCoefficient(v_T, -capacity)
            c_upper.SetCoefficient(v_X, 1)

            T_upper_linking_constraints.append(c_upper)

assert len(T_lower_linking_constraints) == 2190, "The number of constraints (" + str(len(T_lower_linking_constraints)) + ") is NOT correct"
assert len(T_upper_linking_constraints) == 2190, "The number of constraints (" + str(len(T_upper_linking_constraints)) + ") is NOT correct"

#### 1.2. Linking Variable Q

$\sum_{i \in I} X_{i, j, k} - 1 \leq c_{j}Q_{j, k}, \forall j \in J, k \in K$

In [25]:
Q_linking_constraints = []

for j in J:
    capacity = capacities[j]
    for k in K:
        c = solver.Constraint(-solver.infinity(), 1)
        v_Q = Q[(j, k)]
        for i in I:
            v_X = X[(i, j, k)]

            c.SetCoefficient(v_X, 1)
            
        c.SetCoefficient(v_Q, -capacity)
        
        Q_linking_constraints.append(c)

assert len(Q_linking_constraints) == 438, "The number of constraints (" + str(len(Q_linking_constraints)) + ") is NOT correct"

### 2. Station capacity

$\sum_{i \in I}^{}X_{i, j, k} \leq c_{j}, \forall j \in J, k \in K$

In [13]:
response2 = requests.get("https://gitlab.com/drvicsana/cop-proyecto-2023/-/raw/main/located_firehouses.json")
stations = json.loads(response2.text)

In [6]:
# get station capacities
capacities = {}

for station in stations["features"]:
    station_address = station["properties"]["FacilityAddress"]
    capacities[station_address] = station["properties"]["capacity"]

NameError: name 'stations' is not defined

In [15]:
station_capacity_constraints = []

for j in J:
    capacity = capacities[j]
    for k in K:
        c = solver.Constraint(-solver.infinity(), capacity)
    
        for i in I:
            v = X[(i, j, k)]
            c.SetCoefficient(v, 1)

        station_capacity_constraints.append(c)

assert len(station_capacity_constraints) == 438, "The number of constraints (" + str(len(station_capacity_constraints)) + ") is NOT correct"

### 3. Vehicles availability

$\sum_{j \in J}^{}X_{i, j, k} = v_{i}, \forall i \in I, k \in K$

In [16]:
availabilities = {'engine': 197, 'ladder': 143, 'hazardous': 1, 'squad': 8, 'rescue':5}

In [17]:
vehicles_availability_constraints = []

for i in I:
    availability = availabilities[i]

    for k in K:
        c = solver.Constraint(availability, availability)

        for j in J:
            v = X[(i, j, k)]
            c.SetCoefficient(v, 1)
            
        vehicles_availability_constraints.append(c)

assert len(vehicles_availability_constraints) == 10, "The number of constraints (" + str(len(vehicles_availability_constraints)) + ") is NOT correct"

### 4. At least one vehicle

$\sum_{i \in I}^{}X_{i, j, k} \geq 1, \forall j \in J, k \in K$

In [18]:
atleast_one_vehicle_constraints = []

for j in J:
    for k in K:
        c = solver.Constraint(1, solver.infinity())
    
        for i in I:
            v = X[(i, j, k)]
            c.SetCoefficient(v, 1)
        atleast_one_vehicle_constraints.append(c)

assert len(atleast_one_vehicle_constraints) == 438, "The number of constraints (" + str(len(atleast_one_vehicle_constraints)) + ") is NOT correct"

### 5. Scarce resources coverage

$\sum_{j \in J}^{}s_{i, j, n}X_{i, j, k} \geq  1, \forall n \in N, k \in K, i \in \left \{ 3, 4 \right \}$

In [19]:
scarce_resources_availability = []

for k in K:
    for n in N:
        for i in ["rescue", "squad"]:
            c = solver.Constraint(1, solver.infinity())
            
            for j in J:
                s = 1 if serviceability[i].loc[n, j] == True else 0
                v = X[(i, j, k)]
                c.SetCoefficient(v, s)

            scarce_resources_availability.append(c)

assert len(scarce_resources_availability) == 780, "The number of constraints (" + str(len(scarce_resources_availability)) + ") is NOT correct"

### 6. Area resources concentration

$\theta = \sum_{j \in J}^{}s_{i, j, n}$

#### 6.1 Area resource concentration for Engine and Ladder

$\frac{\theta}{2}\leq \sum_{j \in J}^{}s_{i, j, n}X_{i, j, k} \leq \theta, \forall n \in N, k \in K, i \in \left \{ 1, 2 \right \}$

In [20]:
area_resources_concentration_constraints = []

for k in K:
    for n in N:
        for i in ["engine", "ladder"]:
            s_max = 0
            
            for j in J:
                s_max += 1 if serviceability[i].loc[n, j] == True else 0
            
            c = solver.Constraint(s_max / 2, s_max)
            
            for j in J:
                s = 1 if serviceability[i].loc[n, j] == True else 0
                v = X[(i, j, k)]
                c.SetCoefficient(v, s)

            area_resources_concentration_constraints.append(c)
        
assert len(area_resources_concentration_constraints) == 780, "The number of constraints (" + str(len(area_resources_concentration_constraints)) + ") is NOT correct"

#### 6.2 Area resource concentration for Rescue and Squad

$r_{a, b} \geq \frac{l_{i}}{2} \left (T_{i, a, k} + T_{i, b, k} - 1 \right ), \forall k \in K, i \in \left \{ 3, 4 \right \}$

In [21]:
scarce_resources_concentration = []

for k in K:
    for a in J:
        for b in J:
            route_distance = df_distances_stations.loc[a, b]
            for i in ["rescue", "squad"]:
                type_distance_limit = serviceability_distances[i] / 2
                
                c = solver.Constraint(-solver.infinity(), route_distance + type_distance_limit)
            
                T_a = T[(i, a, k)]
                T_b = T[(i, b, k)]

                c.SetCoefficient(T_a, type_distance_limit)
                c.SetCoefficient(T_b, type_distance_limit)

                scarce_resources_concentration.append(c)

assert len(scarce_resources_concentration) == 191844, "The number of constraints (" + str(len(scarce_resources_concentration)) + ") is NOT correct"

### 7. Station resources concentration

#### 7.1 Station resources concentration for Engine and Ladder

$X_{i, j, k} \leq  2, \forall j \in J, k \in K, i \in \left \{ 1, 2 \right \}$

In [22]:
station_resources_concentration_constraints = []

for k in K:
    for j in J:
        for i in ["engine", "ladder"]:
            c = solver.Constraint(-solver.infinity(), 2)
            v = X[(i, j, k)]
            c.SetCoefficient(v, 1)
            station_resources_concentration_constraints.append(c)
                    
assert len(station_resources_concentration_constraints) == 876, "The number of constraints (" + str(len(station_resources_concentration_constraints)) + ") is NOT correct"

#### 7.2 Station resources concentration for Rescue and Squad

$X_{i, j, k} \leq  1, \forall j \in J, k \in K, i \in \left \{ 3, 4 \right \}$

In [23]:
station_resources_concentration_constraints = []

for k in K:
    for j in J:
        for i in ["rescue", "squad"]:
            c = solver.Constraint(-solver.infinity(), 1)
            v = X[(i, j, k)]
            c.SetCoefficient(v, 1)
            station_resources_concentration_constraints.append(c)
                    
assert len(station_resources_concentration_constraints) == 876, "The number of constraints (" + str(len(station_resources_concentration_constraints)) + ") is NOT correct"

### 8. At least two vehicle types

$\sum_{i \in I} T_{i, j, k} \geq 2Q_{j, k}, \forall j \in J, k \in K$

In [26]:
atleast_two_types_of_vehicle_constraints = []

for j in J:
    capacity = capacities[j]
    for k in K:
        c = solver.Constraint(0, solver.infinity())
        v_Q = Q[(j, k)]
        for i in I:
            v_T = T[(i, j, k)]

            c.SetCoefficient(v_T, 1)
        c.SetCoefficient(v_Q, -2)
        atleast_two_types_of_vehicle_constraints.append(c)

assert len(atleast_two_types_of_vehicle_constraints) == 438, "The number of constraints (" + str(len(atleast_two_types_of_vehicle_constraints)) + ") is NOT correct"

### 9. Vehicles transfer

$X_{i, a, 1} - X_{i, a, 2} = \sum_{j \in J} \left ( Y_{i, a, j} - Y_{i, j, a} \right ), \forall i \in I, a \in J$

In [27]:
vehicles_transfer_constraints = []

for i in I:
    for a in J:
        c = solver.Constraint(0, 0)

        for j in J:
            Y_aj = Y[(i, a, j)]
            Y_ja = Y[(i, j, a)]

            c.SetCoefficient(Y_aj, 1)
            c.SetCoefficient(Y_ja, -1)
        
        X_a1 = X[(i, a, 1)]
        X_a2 = X[(i, a, 2)]

        c.SetCoefficient(X_a1, -1)
        c.SetCoefficient(X_a2, 1)

        vehicles_transfer_constraints.append(c)

assert len(vehicles_transfer_constraints) == 1095, "The number of constraints (" + str(len(vehicles_transfer_constraints)) + ") is NOT correct"

### 10. Maximum transfer distance

$r_{a, b}Y_{i, a, b} \leq l_{i} Y_{i, a, b}, \forall i \in \left \{ 1, 2, 3, 4 \right \}, a \in J, b \in J$

In [28]:
max_transfer_distance_constraints = []
a_counter = -1

for a in J:
    for b in J:
        for i in ["engine", "ladder", "rescue", "squad"]:
        
            route_distance = df_distances_stations.loc[a, b]

            #type_distance_limit = serviceability_distances[i] if i in ["engine", "ladder"] else (serviceability_distances[i] / 2)
            type_distance_limit = serviceability_distances[i]

            c = solver.Constraint(-solver.infinity(), 0)
            
            v = Y[(i, a, b)]

            c.SetCoefficient(v, route_distance - type_distance_limit)

            max_transfer_distance_constraints.append(c)

assert len(max_transfer_distance_constraints) == 191844, "The number of constraints (" + str(len(max_transfer_distance_constraints)) + ") is NOT correct"

## Objective Function

$maxZ: \sum_{k \in K}\sum_{i \in I}\sum_{j \in J}X_{j, i, k} \left ( \sum_{n \in N}s_{i, j, n} importance_{j, i, k, n} \right )$

$importance_{j, i, k, n} = w_{p} p^{\varsigma}_{n} + w_{e} e^{\varsigma}_{i, k, n} - w_{d} d^{\varsigma}_{j, n} - v^{\varsigma}_{i}$

In [29]:
MM_scaler = MinMaxScaler()
distance_scaled = MM_scaler.fit_transform(distance)
df_distance_scaled = distance.copy()
df_distance_scaled[:] = distance_scaled
df_distance_scaled.head()

42 South Street  49 Beekman Street  100 Duane Street  \
QN08         0.712966           0.694468          0.704228   
BX28         0.631054           0.599836          0.643624   
QN55         0.623278           0.563582          0.599072   
BK50         0.501529           0.493496          0.528346   
BX41         0.588020           0.555329          0.600535   

      14 N. Moore Street  75 Canal Street  25 Pitt Street  222 East 2 Street  \
QN08            0.707264         0.681702        0.661229           0.645215   
BX28            0.688625         0.648207        0.612963           0.569798   
QN55            0.618132         0.585069        0.568148           0.549079   
BK50            0.550370         0.547323        0.531786           0.511528   
BX41            0.645981         0.601837        0.568300           0.523666   

      340 East 14 Street  253 Lafayette Street  42 Great Jones Street  ...  \
QN08            0.637852              0.689578               0.678506  ...   
BX28            0.538405              0.619114               0.599397  ...   
QN55            0.547624              0.595334               0.585267  ...   
BK50            0.560687              0.558523               0.577703  ...   
BX41            0.494979              0.573894               0.554524  ...   

      256 Hylan Blvd.  278 McClean Avenue  1850 Clove Road  \
QN08         0.646414            0.649381         0.645223   
BX28         0.855937            0.859395         0.835534   
QN55         0.548713            0.554022         0.547603   
BK50         0.436172            0.444186         0.435160   
BX41         0.820951            0.825249         0.789806   

      1592 Richmond Road  256 Nelson Ave.  345 Annadale Road  \
QN08            0.654903         0.810685           0.807546   
BX28            0.841614         0.847833           0.758353   
QN55            0.559543         0.728941           0.724448   
BK50            0.449707         0.634785           0.628735   
BX41            0.796946         0.809545           0.719431   

      1560 Drumgoole Road West  7219 Amboy Road  1100 Rossville Ave  \
QN08                  0.847927         0.872241            0.840141   
BX28                  0.727975         0.709328            0.711457   
QN55                  0.769168         0.788933            0.757351   
BK50                  0.678453         0.692978            0.661992   
BX41                  0.691083         0.670305            0.672679   

      305 Front Street  
QN08          0.663432  
BX28          0.862873  
QN55          0.570430  
BK50          0.463308  
BX41          0.821289  

[5 rows x 219 columns]

In [30]:
availabilities_list = np.array(list(availabilities.values())).reshape(-1, 1)
vehicle_availabilities_scaled = MM_scaler.fit_transform(availabilities_list)
vehicle_availabilities_scaled

array([[1.        ],
       [0.7244898 ],
       [0.        ],
       [0.03571429],
       [0.02040816]])

In [31]:
availabilities_scaled = {'engine': 1.0, 'ladder': 0.7244898, 'hazardous': 0.0, 'squad': 0.03571429, 'rescue':0.02040816}

In [32]:
def set_objective_function(wp, we, wd):
    objective = solver.Objective()
    objective.SetMaximization()


    for k in K:
        shift_str = 'first' if k == 1 else 'second'
        for i in I:
            for j in J:
                v = X[(i, j, k)]
                overall_importance = 0
                
                for n in N:
                    #computing the coefficient
                    area_importance = 0

                    s = 1 if serviceability[i].loc[n][j] else 0

                    if s == 1:  
                        current_n = neighborhood[neighborhood['nta'] == n]
                        column_name = f"{i}_needed_{shift_str}_shift"

                        p = current_n['population_density'].values[0]
                        e = current_n[column_name].values[0]
                        d = df_distance_scaled.loc[n][j]
                        q = availabilities_scaled[i]
                        
                        importance = (wp * p) + (we * e) - (wd * d) - q
                        area_importance += importance

                    overall_importance += area_importance
                objective.SetCoefficient(v, overall_importance)
    
    return objective

## Solve problem

In [33]:
def solve_problem(wp, we, wd):
    deployments_first_shift = pd.DataFrame(0, index=I, columns=J)
    deployments_second_shift = pd.DataFrame(0, index=I, columns=J)

    objective = set_objective_function(wp, we, wd)
    result = solver.Solve()

    if result == solver.ABNORMAL :
        print("Execution finished by an error")
        
    elif result == solver.FEASIBLE :
        print("In the specified time limit the solver has found a feasible solution")
        for k in K:
            for i in I:
                for j in J:
                    v = X[(i,j, k)]
                    if v.SolutionValue()>0:
                        print(v, v.solution_value())
        print("The value for the objective function is", objective.Value())
            
    elif result == solver.INFEASIBLE :
        print("There is no feasible solution for the problem")
            
    elif result == solver.NOT_SOLVED :
        print("In the specified time limit the solver has not found any feasible solution")
            
    elif result == solver.OPTIMAL :
        print("In the specified time limit the solver has found a feasible solution")
        for k in K:
            for i in I:
                for j in J:
                    v = X[(i,j,k)]
                    if v.SolutionValue()>0:
                        if k == 1:
                            deployments_first_shift.loc[i, j] = v.solution_value()
                        else:
                            deployments_second_shift.loc[i, j] = v.solution_value()
                        print(v, v.solution_value())
        
        print("The optimal value for the objective function is", objective.Value())
        
    elif result == solver.UNBOUNDED :
        print("The solution is unbounded")
        
    else :
        print("Unknown error code")

    return deployments_first_shift, deployments_second_shift

In [34]:
wp = 0.40
we = 0.10
wd = 0.50

deployments_first_shift, deployments_second_shift = solve_problem(wp, we, wd)

deployments_first_shift.to_csv("dataset/results/deployments_first_" + str(int(wp*100)) + "p_" + str(int(we*100)) + "e_" + str(int(wd*100)) + "d.csv")
deployments_second_shift.to_csv("dataset/results/deployments_second_" + str(int(wp*100)) + "p_" + str(int(we*100)) + "e_" + str(int(wd*100)) + "d.csv")

In the specified time limit the solver has found a feasible solution
Number of vehicles of type engine deployed at station 42 South Street during shift 1 1.0
Number of vehicles of type engine deployed at station 49 Beekman Street during shift 1 1.0
Number of vehicles of type engine deployed at station 100 Duane Street during shift 1 2.0
Number of vehicles of type engine deployed at station 14 N. Moore Street during shift 1 1.0
Number of vehicles of type engine deployed at station 75 Canal Street during shift 1 1.0
Number of vehicles of type engine deployed at station 25 Pitt Street during shift 1 1.0
Number of vehicles of type engine deployed at station 222 East 2 Street during shift 1 1.0
Number of vehicles of type engine deployed at station 253 Lafayette Street during shift 1 1.0
Number of vehicles of type engine deployed at station 42 Great Jones Street during shift 1 1.0
Number of vehicles of type engine deployed at station 108 East 13 Street during shift 1 1.0
Number of vehicles o